rust-runtime/aws-smithy-runtime-api/src/client/interceptors.rs

```diff
@@ -537,12 +538,50 @@ pub trait Interceptor: std::fmt::Debug {
     );
 }
 
-pub type SharedInterceptor = Arc<dyn Interceptor + Send + Sync>;
+/// Interceptor wrapper that may be shared
+#[derive(Debug, Clone)]
+pub struct SharedInterceptor(Arc<dyn Interceptor + Send + Sync>);
+
+impl SharedInterceptor {
+    /// Create a new `SharedInterceptor` from `Interceptor`
+    pub fn new(interceptor: impl Interceptor + Send + Sync + 'static) -> Self {
+        Self(Arc::new(interceptor))
+    }
+}
+
+impl AsRef<dyn Interceptor> for SharedInterceptor {
+    fn as_ref(&self) -> &(dyn Interceptor + 'static) {
+        self.0.as_ref()
+    }
+}
+
+impl From<Arc<dyn Interceptor + Send + Sync + 'static>> for SharedInterceptor {
+    fn from(interceptor: Arc<dyn Interceptor + Send + Sync + 'static>) -> Self {
+        SharedInterceptor(interceptor)
+    }
+}
+
+impl Deref for SharedInterceptor {
+    type Target = Arc<dyn Interceptor + Send + Sync>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// Collection of [`SharedInterceptor`] that allows for only adding
+#[derive(Debug, Clone, Default)]
+pub struct AddOnlyInterceptors(Vec<SharedInterceptor>);
```

Comment:
Consider calling this `InterceptorRegistrar` since the only function in it is `register`.

Reply:
I like `InterceptorRegistrar` much better. Renamed in 8c1fc01.